AppCenter/AppCenter/Internals/Storage/MSStorageBindableArray.h

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <sqlite3.h>
+
+#import "MSStorageBindableType.h"
+#import <Foundation/Foundation.h>
+
+@interface MSStorageBindableArray : NSObject
+
+/**
+ * Custom array for storing values to be bound in an sqlite statement.
+ * Accepts only supported types.
+ */
+@property(nonatomic) NSMutableArray<id<MSStorageBindableType>> *array;
+
+/**
+ * Adds a string object into array.
+ * @param value NSString value to be added to the array.
+ */
+- (void)addString:(NSString *)value;
+
+/**
+ * Adds a number object into array.
+ * @param value NSNumber value to be added to the array.
+ */
+- (void)addNumber:(NSNumber *)value;
+
+/**
+ * Adds a null value to the array.
+ */
+- (void)addNullValue;
```

Comment:
I think null values could be implicit. for instance addString would be nullable. It probably doesn't make a lot of sense to accept null number though.

Reply:
I'm not sure about this one. Sqlite makes a clear distinction about null, text and number so I though we should, too. Passing nil as string would be implicit, especially if it only works for string, but not number.

Reply:
What do you think @clpolet ?

Reply:
I like your thinking :). But even though `NULL` is an SQLite data type it's not an affinity type https://www.sqlite.org/datatype3.html#type_affinity. From the consumer point of view I am more interesting in the affinity type because sure I care about if the value can be null but I don't care about if it's actually null. I want to persist it as it is knowing that it can be null. So I am more interesting in this API: `- (void)addString:(nullable NSString *)value;` then having to deal with two APIs like: `- (void)addString:(NSString *)value;` and `- (void)addNullValue;`.

From another angle I prefer:
```
[statementValues addString:myValue];
```
to:
```
    if (myValue) {
      [statementValues addString:myValue];
    } else {
      [statementValues addNullValue];
    }
```

Now about NSNumber we can document why it's not nullable. I think in our code a `nil` `NSNumber` is an error case because I don't think we pass any explicit `nil` for an `NSNumber` or it means some conversion may have failed. But I may be wrong. Also, I don't feel strongly about this one.

Reply:
Yeah, from this point of view seems ok